### PR TITLE
fix(github): allow only pull permission for print command

### PR DESCRIPTION
**Pull request by jetersen**
Without configuring platform run options. Not all repos are downloaded due to conditions.

# What does this change

Instead of:
DEBU[0000] Skipping repository since the token does not have push permissions and the run will not fork  repo=jenkinsci/opslevel-plugin
INFO[0000] Running on 0 repositories

You get:
INFO[0000] Running on 1 repositories
INFO[0000] Cloning and running script                    repo=jenkinsci/opslevel-plugin

Since jenkinsci/opslevel-plugin is a fork I have no push permissions and g.Fork always being false because `configureRunPlatform` is never run.

# Notes for the reviewer
_Put any questions or notes for the reviewer here._

# Checklist
- [x] Made sure the PR follows the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines
- [x] Tests if something new is added

### CODE REVIEW

**Comment by lindell** [comment]
Thanks for the PR. The bug is definitely real and needs to be addressed. The `print` command should only require pull permissions. 

Unfortunately, this does more than that. It does add the fork option (as the title suggests), but that option is intentionally kept out of the print command since there are no forks to be considered when running `print`. The fix should be to ensure that only pull permissions are required and nothing else.

**Comment by jetersen** [comment]
Well even though there is no fork. It filters out the repository. So I wonder what the conditional logic should say.

**Comment by lindell** [comment]
It would have to be indicated to each [VersionController](https://github.com/lindell/multi-gitter/tree/master/internal/scm) that push permission is not needed. 

Adding something like "readOnly" and change this 
https://github.com/lindell/multi-gitter/blob/60bbbdf526ce6e87d0f952fdae4858e4c6954952/internal/scm/github/github.go#L146

to
```go
case !g.Fork && !g.ReadOnly && !permissions["push"]:
```

**Comment by jetersen** [comment]
I'll see if I can get around to it in the weekend to add readonly.

**Comment by jetersen** [comment]
Hopefully this weekend!

**Comment by jetersen** [comment]
@lindell I added `flag` that is set programmatically in the `printCmd()` method. To not change the signature of other methods.

File: cmd/platform.go [hunk 1]
```diff
@@ -31,6 +31,10 @@ func configurePlatform(cmd *cobra.Command) {
 	flags.BoolP("include-subgroups", "", false, "Include GitLab subgroups when using the --group flag.")
 	flags.BoolP("ssh-auth", "", false, `Use SSH cloning URL instead of HTTPS + token. This requires that a setup with ssh keys that have access to all repos and that the server is already in known_hosts.`)
 
+	// This is only used by PrintCmd to mark readOnly mode for version control platform
+	flags.Bool("readOnly", false, "If set, This is running in readonly will be read-only.")
+	_ = flags.MarkHidden("readOnly")
```
Comment (jetersen) [inline cmd/platform.go:36]: This is added inside `configurePlatform` which is run for all cmds.
Marking it hidden as this is set programmatically by the print command.

File: cmd/cmd-print.go [hunk 2]
```diff
@@ -48,6 +48,8 @@ func print(cmd *cobra.Command, args []string) error {
 	strOutput, _ := flag.GetString("output")
 	strErrOutput, _ := flag.GetString("error-output")
 
+	_ = flag.Set("readOnly", "true")
```
Comment (jetersen) [inline cmd/cmd-print.go:51]: ignoring the error here because we know `readOnly` flag is available.

File: internal/scm/github/github.go [hunk 3]
```diff
@@ -143,7 +148,7 @@ func (g *Github) GetRepositories(ctx context.Context) ([]scm.Repository, error)
 		case !permissions["pull"]:
 			log.Debug("Skipping repository since the token does not have pull permissions")
 			continue
-		case !g.Fork && !permissions["push"]:
+		case !g.Fork && !g.ReadOnly && !permissions["push"]:
```
Comment (jetersen) [inline internal/scm/github/github.go:151]: testing against 

```
repo:
- jenkinsci/jenkins-infra-test-plugin # a non forked repo
- jenkinsci/opslevel-plugin # a forked repo
```

**Comment by lindell** [comment]
Thanks for the PR @jetersen. 

> @lindell I added flag that is set programmatically in the printCmd() method. To not change the signature of other methods like getVersionController and createGithubClient

I don't think we are able to do this in a good way, since the flags are used to automatically generate documentation and the help text. We do also expose something, that if set by the user, would make multi-gitter not work properly and create confusion. Changing the signature of getVersionController and createGithubClient has to be done in this case, and that is ok 😄 

This logic should also extend to other SCMs, if you are able to add to them as well, that would be appreciated. I can make sure to test it.

**Comment by jetersen** [comment]
I did not see the same permission check in the other SCM

**Comment by lindell** [comment]
> I did not see the same permission check in the other SCM

Indeed that is the case. Nvm that comment

**Comment by jetersen** [comment]
@lindell I believe mark hidden does hide it from help text and usage: See docs: https://github.com/spf13/pflag/blob/85dd5c8bc61cfa382fecd072378089d4e856579d/README.md#hidden-flags

**Comment by jetersen** [comment]
Also sees relatively frequent usage: 
https://cs.github.com/?q=MarkHidden%20language%3AGo&scopeName=All%20repos&scope=
https://github.com/search?q=MarkHidden+language%3Ago&type=code

**Comment by lindell** [comment]
Seems that it hides it from the help text, but we have some other parts that generate docs that do not take it into consideration (yaml documentation in readme).  That could also be fixed to check for hidden flags.

But depending on hidden flags to pass data around (that should not be set by setting the flag) is unfortunately not something I can accept. It does indeed work, but so does a global variable, or environment variables etc.

**Comment by lindell** [comment]
> Also sees relatively frequent usage: https://cs.github.com/?q=MarkHidden%20language%3AGo&scopeName=All%20repos&scope= https://github.com/search?q=MarkHidden+language%3Ago&type=code

MarkHidden should have a lot of legit usecases, but I can't find any place in the top search results where it is simply used as a way to create a hidden flags where it is only set programmatically.

**Comment by jetersen** [comment]
https://github.com/timothysc/clusteradm/blob/01435a84f9b5dee6717df68300458112437c32f9/main.go#L28-L34 ? 🤔

**Comment by jetersen** [comment]
@lindell fixed

**Comment by lindell** [comment]
Thanks 😄 

> If we add more argument passing we should switch to passing a struct.

Totally agree 👍

**Comment by jetersen** [comment]
@lindell Thanks for accepting my pull request! 👏🎉

**Pull request merged**